app/services/discord/sync_roles_service.rb

```diff
@@ -0,0 +1,193 @@
+module Discord
+  class SyncRolesService
+    attr_reader :error_message
+
+    def initialize(school:)
+      @school = school
+      @error_message = ""
+    end
+
+    def sync
+      return false unless sync_ready?
+
+      sync_server_roles
+      true
+    end
+
+    def cached_roles
+      return [] unless sync_ready?
+
+      @cached_roles ||=
+        begin
+          server_discord_role_ids = @server_roles.pluck(:id)
+          discord_roles.map do |role|
+            OpenStruct.new(
+              id: role.id,
+              discord_id: role.discord_id,
+              color_hex: role.color_hex,
+              name: role.name,
+              will_be_deleted: !role.discord_id.in?(server_discord_role_ids)
+            )
+          end
+        end
+    end
+
+    def fetched_roles
+      return [] unless sync_ready?
+
+      @fetched_roles ||=
+        begin
+          discord_roles_ids = discord_roles.pluck(:discord_id)
+
+          @server_roles.each do |server_role|
+            server_role.will_be_added = !server_role.id.in?(discord_roles_ids)
+          end
+        end
+    end
+
+    def deleted_roles?
+      return false unless sync_ready?
+
+      @deleted_roles ||=
+        discord_roles.where.not(discord_id: server_roles.pluck(:id)).exists?
+    end
+
+    def sync_ready?
+      return error t("school_not_configured") unless school_config.configured?
+
+      fetch_roles
+    end
```

Comment:
It's unusual for a method named `sync_ready?` - which to me reads as "_are we ready to sync?"_ - to actually `fetch_roles`. This was surprising behaviour that took me a while to grok. 

Consider replacing the `sync_ready?` public method with a publicized `fetch_ready` method. Just raise errors as required inside `fetch_roles`.